### PR TITLE
Add MQ2 Smoke detection feature.

### PR DESCRIPTION
**Pull request by Nutz95**
This feature is compaible with MQ2 gaz sensor and will shut down any heating element if it detects any gaz or smoke.
It can also warn user using LCD buzzer of displaying error message on the LCD or even making the LCD blink when any hazzard is detected.
An error is also logged into the console output to notify possible 3Rd party software.

New compatible settings have been updated into the Scalar XL 3D printer profile.

I've checked that this is building using Scalar XL settings and that enabling/disabling settings won't break the compilation.

This code was is working with Marlin 1.1.0 RC3 and should work properly with this dev branch.

### CODE REVIEW

File: Gaz_Detector.cpp [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+ int MQSmokeDetector::ReadDigitalPin(int mq_pin)
+ {
+      return  digitalRead(mq_pin);   // read the input pin
+ }
+
+float MQSmokeDetector::Read(int mq_pin)
+{
+  return ResistanceCalculation(analogRead(mq_pin)); 
+}
+
```
Comment (Blue-Marlin) [inline Gaz_Detector.cpp]: analogRead() uses busy wait for ~104 micro seconds.
That's the reason for we do not use this.
FILAMENT_SENSOR is a proper example how to do this.
```
int analogRead(uint8_t pin)
{
        uint8_t low, high;

        if (pin >= 14) pin -= 14; // allow for channel or pin numbers

        // set the analog reference (high two bits of ADMUX) and select the
        // channel (low 4 bits).  this also sets ADLAR (left-adjust result)
        // to 0 (the default).
        ADMUX = (analog_reference << 6) | (pin & 0x07);

        // start the conversion
        sbi(ADCSRA, ADSC);

        // ADSC is cleared when the conversion finishes
        while (bit_is_set(ADCSRA, ADSC));  //!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

        // we have to read ADCL first; doing so locks both ADCL
        // and ADCH until ADCH is read.  reading ADCL second would
        // cause the results of each conversion to be discarded,
        // as ADCL and ADCH would be locked when it completed.
        low  = ADCL;
        high = ADCH;

        // combine the two bytes
        return (high << 8) | low;
}
```

**Comment by Nutz95** [comment]
OK I'm looking at it.
it seems less straight forward.
But i should be able to fit with what is existing.

**Comment by boelle** [comment]
could this PR be rebased against RCbugfix on the main repro? we are going to cut this repro loose as it just adds to the confusion

**Comment by jbrazio** [comment]
The **MarlinDev** repository will soon be deleted, you should first check if the functionality proposed on this PR is already present on the [RCBugfix branch](https://github.com/MarlinFirmware/Marlin/tree/RCBugFix) and if not rebase and reopen this PR at the [MarlinFirmware/Marlin](https://github.com/MarlinFirmware/Marlin) repository against the [RCBugfix branch](https://github.com/MarlinFirmware/Marlin/tree/RCBugFix) for proper followup from the development team.

Thank you for your interest making Marlin better.

**Pull request closed without merging**